# Exercícios: Tratamento dos Dados

#### Carga da Base de Dados Indicadores Criminais RS

In [24]:
import pandas as pd
from pandas import DataFrame

arquivo = './data/df_crimesRS.json'

df_crimes = pd.read_json(arquivo)
df_crimes.head()

index  Municípios  Homicídio  Doloso  Total de vítimas de Homicidio Doloso  \
0      0      ACEGUA                  1                                     1   
1      1  AGUA SANTA                  0                                     0   
2      2       AGUDO                  2                                     2   
3      3   AJURICABA                  1                                     1   
4      4     ALECRIM                  1                                     1   

   Latrocínio  Furtos  Abigeato*  Furto de Veículo  Roubos  Roubo de Veículo  \
0           0      65         18                 1       5                 1   
1           0      31          3                 3       0                 2   
2           0     123         11                 1       5                 0   
3           0      58          5                 2       0                 0   
4           0      85         19                 1       3                 0   

   Estelionato  Delitos Relacionados à Armas e Munições  \
0            4                                        2   
1            2                                        1   
2           19                                       13   
3            8                                        8   
4            8                                       15   

   Entorpecentes - Posse  Entorpecentes - Tráfico  Vítimas de Latrocínio  \
0                      0                        0                      0   
1                      0                        1                      0   
2                     14                       15                      0   
3                      3                        1                      0   
4                      7                        9                      0   

   Vítimas de Lesão Corp. Seg. Morte   ano  
0                                  0  2018  
1                                  0  2018  
2                                  0  2018  
3                                  0  2018  
4                                  0  2018

#### Pivoteamento Longo/Largo e Cálculo de Totais

A base de dados possui a quantidade de crimes registrados para cada ano e município do RS.<br>
Crie um novo DataFrame contendo somente os valores do crime "Homicídio  Doloso" para todos os municípios e anos.

In [25]:
df_homicidios = df_crimes[[ 'Municípios','Homicídio  Doloso','ano']]
df_homicidios

Municípios  Homicídio  Doloso   ano
0                    ACEGUA                  1  2018
1                AGUA SANTA                  0  2018
2                     AGUDO                  2  2018
3                 AJURICABA                  1  2018
4                   ALECRIM                  1  2018
...                     ...                ...   ...
5462  VISTA ALEGRE DO PRATA                  0  2022
5463           VISTA GAUCHA                  0  2022
5464    VITORIA DAS MISSOES                  0  2022
5465              WESTFALIA                  0  2022
5466              XANGRI-LA                  3  2022

[5467 rows x 3 columns]

Modifique o DataFrame de forma que o nome do município seja o índice das linhas e cada ano seja uma coluna, como abaixo:<br>
<pre>ano	        2012	2013	2014	2015	2016	2017	2018	2019	2020	2021	2022
Municípios
Nome
</pre>

In [26]:
df_homicidios = df_homicidios.pivot(index=['Municípios'], columns='ano', values='Homicídio  Doloso')
df_homicidios

ano                    2012  2013  2014  2015  2016  2017  2018  2019  2020  \
Municípios                                                                    
ACEGUA                  0.0   1.0   0.0   0.0   0.0   0.0   1.0   0.0   1.0   
AGUA SANTA              0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
AGUDO                   2.0   0.0   2.0   0.0   0.0   0.0   2.0   0.0   2.0   
AJURICABA               0.0   1.0   0.0   0.0   0.0   0.0   1.0   0.0   0.0   
ALECRIM                 0.0   2.0   0.0   0.0   1.0   1.0   1.0   1.0   1.0   
...                     ...   ...   ...   ...   ...   ...   ...   ...   ...   
VISTA ALEGRE DO PRATA   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
VISTA GAUCHA            0.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
VITORIA DAS MISSOES     0.0   1.0   0.0   0.0   0.0   0.0   0.0   1.0   0.0   
WESTFALIA               0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
XANGRI-LA               3.0   3.0   3.0   5.0   1.0   4.0   4.0   0.0   4.0   

ano                    2021  2022  
Municípios                         
ACEGUA                  1.0   1.0  
AGUA SANTA              2.0   0.0  
AGUDO                   2.0   1.0  
AJURICABA               0.0   0.0  
ALECRIM                 1.0   1.0  
...                     ...   ...  
VISTA ALEGRE DO PRATA   0.0   0.0  
VISTA GAUCHA            0.0   0.0  
VITORIA DAS MISSOES     1.0   0.0  
WESTFALIA               0.0   0.0  
XANGRI-LA               4.0   3.0  

[498 rows x 11 columns]

Acrescente uma coluna denominada "Total/Município" e calcule o total de homicídios para cada município.

In [27]:
df_homicidios['Total/Município'] = df_homicidios.sum(axis=1)

Acrescente uma linha  denominada "Total/Ano" e calcule o total de homicídios por ano para todos os municípios.

In [28]:
total_ano = DataFrame(df_homicidios.sum(axis=0),columns=['Total/Ano']).T
df_homicidios = pd.concat([df_homicidios,total_ano])
df_homicidios

ano                    2012    2013    2014    2015    2016    2017    2018  \
ACEGUA                  0.0     1.0     0.0     0.0     0.0     0.0     1.0   
AGUA SANTA              0.0     0.0     0.0     0.0     0.0     0.0     0.0   
AGUDO                   2.0     0.0     2.0     0.0     0.0     0.0     2.0   
AJURICABA               0.0     1.0     0.0     0.0     0.0     0.0     1.0   
ALECRIM                 0.0     2.0     0.0     0.0     1.0     1.0     1.0   
...                     ...     ...     ...     ...     ...     ...     ...   
VISTA GAUCHA            0.0     1.0     0.0     0.0     0.0     0.0     0.0   
VITORIA DAS MISSOES     0.0     1.0     0.0     0.0     0.0     0.0     0.0   
WESTFALIA               0.0     0.0     0.0     0.0     0.0     0.0     0.0   
XANGRI-LA               3.0     3.0     3.0     5.0     1.0     4.0     4.0   
Total/Ano            2026.0  1941.0  2377.0  2479.0  2666.0  2706.0  2156.0   

ano                    2019    2020    2021    2022  Total/Município  
ACEGUA                  0.0     1.0     1.0     1.0              5.0  
AGUA SANTA              0.0     0.0     2.0     0.0              2.0  
AGUDO                   0.0     2.0     2.0     1.0             11.0  
AJURICABA               0.0     0.0     0.0     0.0              2.0  
ALECRIM                 1.0     1.0     1.0     1.0              9.0  
...                     ...     ...     ...     ...              ...  
VISTA GAUCHA            0.0     0.0     0.0     0.0              1.0  
VITORIA DAS MISSOES     1.0     0.0     1.0     0.0              3.0  
WESTFALIA               0.0     0.0     0.0     0.0              0.0  
XANGRI-LA               0.0     4.0     4.0     3.0             34.0  
Total/Ano            1700.0  1726.0  1529.0  1623.0          22929.0  

[499 rows x 12 columns]

Liste as linhas contendo dados dos 10 municípios com maior quantidade de homicídios no período de 2012 a 2022, seguida da linha "Total/Ano".

In [29]:
total_municipio = df_homicidios.sort_values('Total/Município',ascending=False)[0:1]
df_top10 = pd.concat([df_homicidios.sort_values('Total/Município',ascending=False)[1:11],total_municipio])
df_top10

ano              2012    2013    2014    2015    2016    2017    2018    2019  \
PORTO ALEGRE    473.0   463.0   583.0   607.0   723.0   594.0   466.0   292.0   
ALVORADA        132.0    95.0   157.0   121.0   107.0   173.0   132.0    85.0   
VIAMAO          106.0   100.0   104.0   114.0   136.0   120.0   115.0    70.0   
CANOAS          120.0   105.0   106.0   138.0   113.0   122.0   110.0    60.0   
CAXIAS DO SUL   107.0    86.0    89.0    89.0   115.0    83.0    78.0    59.0   
GRAVATAI         61.0    73.0    92.0    78.0   101.0   136.0    65.0    67.0   
SAO LEOPOLDO     70.0    57.0   113.0    97.0    76.0    88.0    41.0    43.0   
PELOTAS          55.0    53.0    64.0   101.0    65.0   107.0    68.0    54.0   
NOVO HAMBURGO    87.0    63.0    87.0    80.0    54.0    59.0    43.0    39.0   
SANTA MARIA      28.0    29.0    42.0    52.0    57.0    61.0    56.0    37.0   
Total/Ano      2026.0  1941.0  2377.0  2479.0  2666.0  2706.0  2156.0  1700.0   

ano              2020    2021    2022  Total/Município  
PORTO ALEGRE    260.0   226.0   298.0           4985.0  
ALVORADA        109.0    70.0    63.0           1244.0  
VIAMAO           87.0    59.0    55.0           1066.0  
CANOAS           51.0    54.0    73.0           1052.0  
CAXIAS DO SUL    73.0    81.0    62.0            922.0  
GRAVATAI         51.0    32.0    29.0            785.0  
SAO LEOPOLDO     42.0    32.0    29.0            688.0  
PELOTAS          26.0    23.0    18.0            634.0  
NOVO HAMBURGO    30.0    26.0    23.0            591.0  
SANTA MARIA      42.0    51.0    54.0            509.0  
Total/Ano      1726.0  1529.0  1623.0          22929.0

Calcule o valor percentual de cada um dos municípios em relação ao total, utilizando a linha "Total/Ano" como base.

In [30]:
df_top10.div(df_top10.loc['Total/Ano']/100)[0:10]

ano                 2012       2013       2014       2015       2016  \
PORTO ALEGRE   23.346496  23.853684  24.526714  24.485680  27.119280   
ALVORADA        6.515301   4.894384   6.604964   4.881000   4.013503   
VIAMAO          5.231984   5.151984   4.375263   4.598628   5.101275   
CANOAS          5.923001   5.409583   4.459403   5.566761   4.238560   
CAXIAS DO SUL   5.281343   4.430706   3.744215   3.590157   4.313578   
GRAVATAI        3.010859   3.760948   3.870425   3.146430   3.788447   
SAO LEOPOLDO    3.455084   2.936631   4.753891   3.912868   2.850713   
PELOTAS         2.714709   2.730551   2.692469   4.074223   2.438110   
NOVO HAMBURGO   4.294176   3.245750   3.660076   3.227108   2.025506   
SANTA MARIA     1.382034   1.494075   1.766933   2.097620   2.138035   

ano                 2017       2018       2019       2020       2021  \
PORTO ALEGRE   21.951220  21.614100  17.176471  15.063731  14.780903   
ALVORADA        6.393200   6.122449   5.000000   6.315180   4.578156   
VIAMAO          4.434590   5.333952   4.117647   5.040556   3.858731   
CANOAS          4.508500   5.102041   3.529412   2.954809   3.531720   
CAXIAS DO SUL   3.067258   3.617811   3.470588   4.229432   5.297580   
GRAVATAI        5.025868   3.014842   3.941176   2.954809   2.092871   
SAO LEOPOLDO    3.252033   1.901670   2.529412   2.433372   2.092871   
PELOTAS         3.954176   3.153989   3.176471   1.506373   1.504251   
NOVO HAMBURGO   2.180340   1.994434   2.294118   1.738123   1.700458   
SANTA MARIA     2.254250   2.597403   2.176471   2.433372   3.335513   

ano                 2022  Total/Município  
PORTO ALEGRE   18.361060        21.741027  
ALVORADA        3.881701         5.425444  
VIAMAO          3.388786         4.649134  
CANOAS          4.497843         4.588076  
CAXIAS DO SUL   3.820086         4.021109  
GRAVATAI        1.786815         3.423612  
SAO LEOPOLDO    1.786815         3.000567  
PELOTAS         1.109057         2.765057  
NOVO HAMBURGO   1.417129         2.577522  
SANTA MARIA     3.327172         2.219896

#### Pivoteamento Largo/Longo - Voltando ao Formato Original

Utilize o DataFrame em formato largo, em que foi acrescentada uma coluna e linha de totais para o exercício a seguir.<br>
O objetivo é retornar os dados ao formato original.

Primeiramente, remova do DataFrame a Coluna e a linha de Total.

In [31]:
df_largo = df_homicidios.drop('Total/Município',axis=1)[:-1]
df_largo

ano                    2012  2013  2014  2015  2016  2017  2018  2019  2020  \
ACEGUA                  0.0   1.0   0.0   0.0   0.0   0.0   1.0   0.0   1.0   
AGUA SANTA              0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
AGUDO                   2.0   0.0   2.0   0.0   0.0   0.0   2.0   0.0   2.0   
AJURICABA               0.0   1.0   0.0   0.0   0.0   0.0   1.0   0.0   0.0   
ALECRIM                 0.0   2.0   0.0   0.0   1.0   1.0   1.0   1.0   1.0   
...                     ...   ...   ...   ...   ...   ...   ...   ...   ...   
VISTA ALEGRE DO PRATA   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
VISTA GAUCHA            0.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
VITORIA DAS MISSOES     0.0   1.0   0.0   0.0   0.0   0.0   0.0   1.0   0.0   
WESTFALIA               0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
XANGRI-LA               3.0   3.0   3.0   5.0   1.0   4.0   4.0   0.0   4.0   

ano                    2021  2022  
ACEGUA                  1.0   1.0  
AGUA SANTA              2.0   0.0  
AGUDO                   2.0   1.0  
AJURICABA               0.0   0.0  
ALECRIM                 1.0   1.0  
...                     ...   ...  
VISTA ALEGRE DO PRATA   0.0   0.0  
VISTA GAUCHA            0.0   0.0  
VITORIA DAS MISSOES     1.0   0.0  
WESTFALIA               0.0   0.0  
XANGRI-LA               4.0   3.0  

[498 rows x 11 columns]

Em seguida, utilize a função melt() para transformar do formato largo:
<pre>ano	        2012	2013	2014	2015	2016	2017	2018	2019	2020	2021	2022
Municípios
Nome
</pre>
Para o formato longo:
<pre>
Municípios      Ano     Homicídio Doloso
Nome            2012    123
Nome            2013    150
Nome            2013    160
</pre>
Se necessário, consulte a documentação do método pd.melt() em https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.melt.html<br>
Lembre-se de designar a coluna "Municípios" como índice do DataFrame resultante.

In [32]:
df_longo = df_largo.reset_index().rename(columns={'index':'Municípios'})
df_longo = pd.melt(df_longo,id_vars=['Municípios'],var_name=['ano'],value_name='Homicídio Doloso').dropna()
df_longo.set_index('Municípios')

ano  Homicídio Doloso
Municípios                                   
ACEGUA                 2012               0.0
AGUA SANTA             2012               0.0
AGUDO                  2012               2.0
AJURICABA              2012               0.0
ALECRIM                2012               0.0
...                     ...               ...
VISTA ALEGRE DO PRATA  2022               0.0
VISTA GAUCHA           2022               0.0
VITORIA DAS MISSOES    2022               0.0
WESTFALIA              2022               0.0
XANGRI-LA              2022               3.0

[5467 rows x 2 columns]

#### Carga Base de Dados: População Municípios 2010/2020

In [33]:
df_populacao = pd.read_csv('./data/populacao_municipios.csv',sep=';')
df_populacao.head()

Sigla   Código     Município     2010     2022
0    AC  1200013    Acrelândia  12538.0  14021.0
1    AC  1200054  Assis Brasil   6072.0   8100.0
2    AC  1200104     Brasiléia  21398.0  26000.0
3    AC  1200138        Bujari   8471.0  12917.0
4    AC  1200179      Capixaba   8798.0  10392.0

Vamos calcular a taxa de homicídios por habitante para cada um dos municípios do Rio Grande do Sul.<br>
Para isso, será necessário juntar os DataFrames df_populacao e df_homicidio utilizando o campo "Município" como chave.

#### Preparação dos Dados

Vez que o campo Código não está disponível na base de homicídios ,teremos que fazer uma junção utilizando o nome do município como chave.<br>
Note que, na primeira base, o nome dos municípios está em caixa alta e sem acentos, como "AGUA SANTA".<br>
Na segunda base, os nomes estão grafados da forma usual, como "Água Santa'.<br>
Utilize a seguinte função auxiliar limpa_texto() para modificar os valores da coluna Município do DataFrame df_populacao:

In [34]:
import unicodedata

def limpa_texto(texto='',*kword):
    nfkd_form = unicodedata.normalize('NFKD', texto)
    only_ascii = nfkd_form.encode('ASCII', 'ignore').decode("utf-8")
    texto = only_ascii.upper()
    return texto

In [35]:
df_populacao['Município']=df_populacao['Município'].apply(limpa_texto)
df_populacao

Sigla   Código       Município     2010     2022
0       AC  1200013      ACRELANDIA  12538.0  14021.0
1       AC  1200054    ASSIS BRASIL   6072.0   8100.0
2       AC  1200104       BRASILEIA  21398.0  26000.0
3       AC  1200138          BUJARI   8471.0  12917.0
4       AC  1200179        CAPIXABA   8798.0  10392.0
...    ...      ...             ...      ...      ...
5591    TO  1721208  TOCANTINOPOLIS  22619.0  22615.0
5592    TO  1721257        TUPIRAMA   1574.0   1909.0
5593    TO  1721307      TUPIRATINS   2097.0   1874.0
5594    TO  1722081    WANDERLANDIA  10981.0  10522.0
5595    TO  1722107         XAMBIOA  11484.0  10517.0

[5596 rows x 5 columns]

A base de dados contém dados de todos os estados, o código do município e a população nos anos de 2010 e 2022.<br>
Crie um novo DataFrame contendo somente os dados do Estado do RS, código e a população do ano de 2022.<br>
Modifique o nome das colunas de '2022' para 'População' e "Município' para 'Municípios'.<br>
Utilize o nome do município, como índice das linhas.<br>

In [36]:
df_rs = df_populacao[df_populacao['Sigla']=='RS'][['Município','Código','2022']]
df_rs = df_rs.rename(columns={'Município':'Municípios','2022':'População'})
df_rs = df_rs.set_index('Municípios')

Faça a junção dos DataFrames de população de homicídios e população dos municípios do RS por meio do índice.<br>
Se necessário consulte a documentação do método pd.merge() disponível em: https://pandas.pydata.org/docs/reference/api/pandas.merge.html<br>
Verifique os parâmetros lef_index e right_index.

In [37]:
df_homicidios = pd.merge(df_homicidios, df_rs, left_index=True, right_index=True)
df_homicidios

2012  2013  2014  2015  2016  2017  2018  2019  2020  \
ACEGUA                  0.0   1.0   0.0   0.0   0.0   0.0   1.0   0.0   1.0   
AGUA SANTA              0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
AGUDO                   2.0   0.0   2.0   0.0   0.0   0.0   2.0   0.0   2.0   
AJURICABA               0.0   1.0   0.0   0.0   0.0   0.0   1.0   0.0   0.0   
ALECRIM                 0.0   2.0   0.0   0.0   1.0   1.0   1.0   1.0   1.0   
...                     ...   ...   ...   ...   ...   ...   ...   ...   ...   
VISTA ALEGRE DO PRATA   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
VISTA GAUCHA            0.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
VITORIA DAS MISSOES     0.0   1.0   0.0   0.0   0.0   0.0   0.0   1.0   0.0   
WESTFALIA               0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
XANGRI-LA               3.0   3.0   3.0   5.0   1.0   4.0   4.0   0.0   4.0   

                       2021  2022  Total/Município   Código  População  
ACEGUA                  1.0   1.0              5.0  4300034     4170.0  
AGUA SANTA              2.0   0.0              2.0  4300059     3912.0  
AGUDO                   2.0   1.0             11.0  4300109    16039.0  
AJURICABA               0.0   0.0              2.0  4300208     6720.0  
ALECRIM                 1.0   1.0              9.0  4300307     6123.0  
...                     ...   ...              ...      ...        ...  
VISTA ALEGRE DO PRATA   0.0   0.0              0.0  4323606     1590.0  
VISTA GAUCHA            0.0   0.0              1.0  4323705     2783.0  
VITORIA DAS MISSOES     1.0   0.0              3.0  4323754     3260.0  
WESTFALIA               0.0   0.0              0.0  4323770     3098.0  
XANGRI-LA               4.0   3.0             34.0  4323804    16463.0  

[494 rows x 14 columns]

Armazene os dados do DataFrame em um arquivo JSON para utilização nas próximas aulas.

In [38]:
df_homicidios.to_json('df_homicidios_populacao.json')